src/rgw/rgw_lc.cc

```diff
@@ -1390,3 +1396,132 @@ int RGWLC::remove_bucket_config(RGWBucketInfo& bucket_info,
   return ret;
 }
 
+std::string rgwlc_s3_expiration_header(
+  DoutPrefixProvider* dpp,
+  const rgw_obj_key& obj_key,
+  const RGWObjTags& obj_tagset,
+  const ceph::real_time& mtime,
+  const std::map<std::string, buffer::list>& bucket_attrs)
+{
+  CephContext* cct = dpp->get_cct();
+  RGWLifecycleConfiguration config(cct); // TODO: save in bucket info
+  std::string hdr{""};
+
+  const auto& aiter = bucket_attrs.find(RGW_ATTR_LC);
+  if (aiter == bucket_attrs.end())
+    return hdr;
+
+  bufferlist::const_iterator iter{&aiter->second};
+  try {
+      config.decode(iter);
+  } catch (const buffer::error& e) {
+      ldpp_dout(dpp, 0) << __func__
+			<<  "() decode life cycle config failed"
+			<< dendl;
+      return hdr;
+  } /* catch */
+
+  boost::optional<ceph::real_time> expiration_date;
+  boost::optional<std::string> rule_id;
+
+  const auto& rule_map = config.get_rule_map();
+  for (const auto& ri : rule_map) {
+    const auto& rule = ri.second;
+    auto& id = rule.get_id();
+    auto& prefix = rule.get_prefix();
+    auto& filter = rule.get_filter();
+    auto& expiration = rule.get_expiration();
+    auto& noncur_expiration = rule.get_noncur_expiration();
+
+    ldpp_dout(dpp, 0) << "rule: " << ri.first
+		      << " prefix: " << prefix
+		      << " expiration: "
+		      << " date: " << expiration.get_date()
+		      << " days: " << expiration.get_days()
+		      << " noncur_expiration: "
+		      << " date: " << noncur_expiration.get_date()
+		      << " days: " << noncur_expiration.get_days()
+		      << dendl;
+
+    /* skip if rule !enabled
+     * if rule has prefix, skip iff object !match prefix
+     * if rule has tags, skip iff object !match tags
+     * note if object is current or non-current, compare accordingly
+     * if rule has days, construct date expression and save iff older
+     * than last saved
+     * if rule has date, convert date expression and save iff older
+     * than last saved
+     * if the date accum has a value, format it into hdr
+     */
+
+    if (! rule.is_enabled())
+      continue;
+
+    if(! prefix.empty()) {
+      if (obj_key.name.find(prefix) == string::npos)
```

Comment:
`!= 0`?

Reply:
don't follow

Reply:
@mattbenjamin it's prefix, need to make sure that it's not found at offest 0, not if not found.

Reply:
probably a compare would be more efficient

Reply:
ah, of course, sorry